superset/assets/src/explore/components/controls/DateFilterControl.jsx

```diff
@@ -28,172 +51,265 @@ const propTypes = {
+      isFreeform: { ...this.state.isFreeform },
+    };
+    if (value === INVALID_DATE_MESSAGE) {
+      // the DateTimeField component will return `Invalid date` for freeform
+      // text, so we need to cheat and steal the value from the state
+      const freeformValue = this.dateTimeFieldRefs[key].state.inputValue;
+      nextState.isFreeform[key] = true;
+      nextState[key] = freeformValue;
+    } else {
+      nextState.isFreeform[key] = false;
+      nextState[key] = value;
+    }
+    this.setState(nextState, this.updateRefs);
   }
-  setType(type) {
-    this.setState({ type });
+  setCommonRange(timeFrame) {
+    const nextState = {
+      type: 'range',
+      common: timeFrame,
+      until: moment().startOf('day').format(MOMENT_FORMAT),
+    };
+    let units;
+    if (timeFrame === 'Yesterday') {
+      units = 'days';
+    } else {
+      units = timeFrame.split(' ')[1] + 's';
+    }
+    nextState.since = moment().startOf('day').subtract(1, units).format(MOMENT_FORMAT);
+    this.setState(nextState, this.updateRefs);
   }
-  setValueAndClose(val) {
-    this.setState({ type: 'free', free: val }, this.close);
+  setCustomRange(key, value) {
+    const nextState = { ...this.state, type: 'range', common: null };
+    if (key !== undefined && value !== undefined) {
+      nextState[key] = value;
+    }
+    if (nextState.rel === 'Last') {
+      nextState.until = moment().startOf('day').format(MOMENT_FORMAT);
+      nextState.since = moment()
+        .startOf('day')
+        .subtract(nextState.num, nextState.grain)
+        .format(MOMENT_FORMAT);
+    } else {
+      nextState.until = moment()
+        .startOf('day')
+        .add(nextState.num, nextState.grain)
+        .format(MOMENT_FORMAT);
+      nextState.since = moment().startOf('day').format(MOMENT_FORMAT);
+    }
+    this.setState(nextState, this.updateRefs);
   }
-  setDatetime(dttm) {
-    this.setState({ dttm: dttm.format().substring(0, 19) });
+  updateRefs() {
+    this.dateTimeFieldRefs.since.setState({ inputValue: this.state.since });
+    this.dateTimeFieldRefs.until.setState({ inputValue: this.state.until });
   }
   close() {
     let val;
-    if (this.state.type === 'rel') {
-      val = `${this.state.num} ${this.state.grain} ${this.state.rel}`;
-    } else if (this.state.type === 'fix') {
-      val = this.state.dttm;
-    } else if (this.state.type === 'free') {
-      val = this.state.free;
+    if (this.state.type === 'range') {
+      val = this.state.common
+        ? this.state.common
+        : `${this.state.rel} ${this.state.num} ${this.state.grain}`;
+    } else if (this.state.type === 'startend') {
+      val = [this.state.since, this.state.until].join(SEPARATOR);
     }
     this.props.onChange(val);
     this.refs.trigger.hide();
   }
   renderPopover() {
+    const grainOptions = TIME_GRAIN_OPTIONS.map((grain, index) => (
+      <MenuItem
+        onSelect={this.setCustomRange.bind(this, 'grain')}
+        key={'grain-' + index}
```

Comment:
generally using `index` as a key is considered an antipattern (https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318)

also style nit: I don't think you need to add that `grain-` prefix

Reply:
Ah, #TIL. Will fix, though in this case it should be safe (no reordering or updates in the list).